aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -1569,6 +1569,361 @@ std::tuple<Tensor&, Tensor&> symeig_out(Tensor& vals, Tensor& vecs, const Tensor
+    } else {
+      TORCH_CHECK(false, "linalg_eig: imaginary part of eigenvalues is non-zero, can't safely cast eigenvalues to non-complex dtype.")
+    }
+    if (compute_eigenvectors) {
+      if (vectors.is_complex()) {
+          vectors = linalg_eig_make_complex_eigenvectors(vectors, values, maybe_complex_vectors);
+      } else {
+        TORCH_CHECK(false, "linalg_eig: imaginary part of eigenvectors is non-zero, can't safely cast eigenvectors to non-complex dtype.")
+      }
+    }
+  }
+
+  return std::tuple<Tensor&, Tensor&>(values, vectors);
+}
+
+std::tuple<Tensor&, Tensor&> linalg_eig_out(const Tensor& input, Tensor& values, Tensor& vectors) {
+  squareCheckInputs(input);
+
+  checkLinalgCompatibleDtype("linalg_eig", values, input, "eigenvalues");
+  checkLinalgCompatibleDtype("linalg_eig", vectors, input, "eigenvectors");
+
+  // MAGMA doesn't have GPU interface for GEEV routine, it requires inputs to be on CPU
+  auto options = input.options().device(at::kCPU);
+  auto infos = at::zeros({std::max<int64_t>(1, batchCount(input))}, options.dtype(kInt));
+
+  // if result is not empty and not in batched column major format we have to allocate a temporary tensor
+  bool is_batched_column_major = false;
+  if (vectors.dim() >= 2) {
+    is_batched_column_major = vectors.transpose(-2, -1).is_contiguous();
+  }
+
+  // for NumPy compatibility we can't determine the dtype of eigenvalues before running the computation
+  // for real-valued inputs NumPy can return either real- or complex-valued output, we will create values_tmp of complex type
+  // for complex-valued inputs we can use values' storage directly if dtype matches
+  // bool values_input_same_type = input.is_complex() ? (values.scalar_type() == input.scalar_type()) : false;
+  bool values_expected_type = (values.scalar_type() == toComplexType(input.scalar_type()));
+  bool vectors_expected_type = (vectors.scalar_type() == toComplexType(input.scalar_type()));
+
+  auto expected_values_shape = IntArrayRef(input.sizes().data(), input.dim()-1);  // input.shape[:-1]
+  bool values_equal_expected_shape = values.sizes().equals(expected_values_shape);
+  bool vectors_equal_expected_shape = vectors.sizes().equals(input.sizes());
+
+  // if result is not empty and not in batched column major format
+  bool values_tmp_needed = (values.numel() != 0 && !values.is_contiguous());
+  bool vectors_tmp_needed = (vectors.numel() != 0 && !is_batched_column_major);
+  // or result does not have the expected shape
+  values_tmp_needed |= (values.numel() != 0 && !values_equal_expected_shape);
+  vectors_tmp_needed |= (vectors.numel() != 0 && !vectors_equal_expected_shape);
+  // or result does not have the expected dtype
+  values_tmp_needed |= !values_expected_type;
+  vectors_tmp_needed |= !vectors_expected_type;
+  // we will allocate a temporary tensor and do the copy
+
+  // "out=" variants shouldn't allow result and input to be on different devices
```

Comment:
https://github.com/pytorch/pytorch/pull/53259 will add checks for this to all OpInfos

Reply:
This function is special. MAGMA takes CPU inputs and return CPU outputs and internally calls GPU for some parts of the computation managing CPU-GPU data transfers. If the passed tensors to out= live on CPU while the input is on GPU, then it's actually a beneficial situation involving fewer explicit CPU-GPU copies. Should we be consistent and disallow out= and input to be on different devices?

Reply:
Yes. I appreciate your reasoning and you're completely right, but supporting these little oddities has some cost and in the future we may be able to switch to a different implementation that is actually run on the GPU, and then our oddity will become a liability.